src/dmd/func.d

```diff
@@ -4271,3 +4302,41 @@ extern (C++) final class NewDeclaration : FuncDeclaration
         v.visit(this);
     }
 }
+
+/// Stores a reason why a function was inferred `@system`
+///
+/// Has two modes:
+/// - a regular safety error, stored in (fmtStr, arg0, arg1)
+/// - a call to a `@system` function, which is a special case, because in that case,
+///   that function might recursively also have a `SafetyViolation`. This way, in case
+///   of a big call stack, the error can go down all the way to the root cause.
+///   The `FunctionDeclaration` is then stored in `arg0` and `fmtStr` must be `null`.
+private struct SafetyViolation
+{
+    /// location of error
+    Loc loc = Loc.init;
+    /// printf-style format string
+    const(char)* fmtStr = null;
+    /// Arguments for up to two `%s` format specifiers in format string
+    RootObject arg0 = null;
+    /// ditto
+    RootObject arg1 = null;
+}
+
+/// Print the reason why `fd` was inferred `@system` as a supplemental error
+void errorSupplementalInferredSafety(FuncDeclaration fd)
+{
+    if (auto s = fd.safetyViolation)
+    {
+        if (s.fmtStr)
+        {
+            errorSupplemental(s.loc, "which was inferred `@system` because of:");
```

Comment:
We could present a message when error number limit is exceeded by suggesting `-v` to be more verbose. This makes the error consistent with other errors like template overloading mismatch.

Reply:
Yes, I was thinking of adding a depth limit anyway to avoid endless recursion (currently shouldn't be a problem, but when circular dependencies in attribute inference are fixed it could pop up)